File: client/lib/models/room.dart
class RoomState {
  final String code;
  final String hostPlayerId;
  final int maxPlayers;
  final String status;
  final List<RoomPlayer> players;

  RoomState({
    required this.code,
    required this.hostPlayerId,
    required this.maxPlayers,
    required this.status,
    required this.players,
  });

  factory RoomState.fromJson(Map<String, dynamic> json) => RoomState(
    code: json['code'] as String,
    hostPlayerId: json['hostPlayerId'] as String,
    maxPlayers: json['maxPlayers'] as int,
    status: json['status'] as String,
    players: (json['players'] as List)
        .map((p) => RoomPlayer.fromJson(p as Map<String, dynamic>))
        .toList(),
  );

  bool get isHost => true; // determined by provider
  bool get canStart => players.length >= 3 && status == 'waiting';
}

class RoomPlayer {
  final String id;
  final String name;
  final int seatIndex;
  final bool isConnected;
  final bool isAI;

  RoomPlayer({
    required this.id,
    required this.name,
    required this.seatIndex,
    required this.isConnected,
    required this.isAI,
  });

  factory RoomPlayer.fromJson(Map<String, dynamic> json) => RoomPlayer(
    id: json['id'] as String,
    name: json['name'] as String,
    seatIndex: json['seatIndex'] as int,
    isConnected: json['isConnected'] as bool,
    isAI: json['isAI'] as bool? ?? false,
  );
}

File: server/src/room/room.ts
import { RoomState, RoomPlayerInfo } from '../types/events';

export interface RoomPlayer {
  id: string;
  name: string;
  socketId: string;
  seatIndex: number;
  isConnected: boolean;
  isAI: boolean;
}

export class Room {
  code: string;
  hostPlayerId: string;
  maxPlayers: number;
  status: 'waiting' | 'playing' | 'finished';
  players: RoomPlayer[];
  createdAt: number;
  lastActivityAt: number;

  constructor(code: string, maxPlayers: number) {
    this.code = code;
    this.maxPlayers = maxPlayers;
    this.status = 'waiting';
    this.players = [];
    this.createdAt = Date.now();
    this.lastActivityAt = Date.now();
    this.hostPlayerId = '';
  }

  addPlayer(player: RoomPlayer): void {
    if (this.players.length === 0) {
      this.hostPlayerId = player.id;
    }
    this.players.push(player);
    this.lastActivityAt = Date.now();
  }

  removePlayer(playerId: string): RoomPlayer | null {
    const index = this.players.findIndex((p) => p.id === playerId);
    if (index === -1) return null;

    const removed = this.players.splice(index, 1)[0];

    if (this.players.length > 0 && this.hostPlayerId === playerId) {
      this.hostPlayerId = this.players[0].id;
    }

    this.lastActivityAt = Date.now();
    return removed;
  }

  getPlayer(playerId: string): RoomPlayer | undefined {
    return this.players.find((p) => p.id === playerId);
  }

  isFull(): boolean {
    return this.players.length >= this.maxPlayers;
  }

  isEmpty(): boolean {
    return this.players.length === 0;
  }

  allConnected(): boolean {
    return this.players.every((p) => p.isConnected || p.isAI);
  }

  getHumanPlayers(): RoomPlayer[] {
    return this.players.filter((p) => !p.isAI);
  }

  toState(): RoomState {
    return {
      code: this.code,
      hostPlayerId: this.hostPlayerId,
      maxPlayers: this.maxPlayers,
      status: this.status,
      players: this.players.map((p) => ({
        id: p.id,
        name: p.name,
        seatIndex: p.seatIndex,
        isConnected: p.isConnected,
        isAI: p.isAI,
      })),
    };
  }
}
